### PR TITLE
Include Elasticsearch in the load tester cluster

### PR DESCRIPTION
**Pull request by ethervoid**
#### Summary

This PR adds Elasticsearch as an optional part for the load tester. It will include:

- Elasticsearch instance configured
- Elasticsearch Prometheus metrics exporter
- New Elasticsearch dasboard

Screenshot from 2020-06-02 21-31-02

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-25269

### CODE REVIEW

File: config/deployer.default.json [hunk 1]
```diff
@@ -1,9 +1,17 @@
 {
   "ClusterName": "loadtest",
+  "VpcId": "",
```
Comment (ethervoid) [inline config/deployer.default.json:3]: We can probably use this id and put all the machines in the same VPC

File: deployment/config.go [hunk 2]
```diff
@@ -48,6 +50,14 @@ type Config struct {
 	LogSettings         logger.Settings
 }
 
+type ElasticSearchSettings struct {
```
Comment (agnivade) [inline deployment/config.go:53]: godocs please :)

File: deployment/config.go [hunk 3]
```diff
@@ -48,6 +50,14 @@ type Config struct {
...
+	Enable       bool   // Elasticsearch is going to be created
```
Comment (agnivade) [inline deployment/config.go:54]: This is a very small thing. So feel free to ignore. But all comments here end with a period. So might be good to keep that consistency.

File: deployment/terraform/create.go [hunk 4]
```diff
@@ -338,6 +344,16 @@ func (t *Terraform) updateAppConfig(ip string, sshc *ssh.Client, output *Output)
 	cfg.PluginSettings.Enable = model.NewBool(true)
 	cfg.PluginSettings.EnableUploads = model.NewBool(true)
 
+	if t.config.HasElasticSearch() {
```
Comment (agnivade) [inline deployment/terraform/create.go:347]: We usually check the output struct for this. It helps confirm that Value[0] is actually there. We can add a similar method to the output struct also like `HasProxy`.

File: deployment/terraform/destroy.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ func (t *Terraform) Destroy() error {
 		"-var", fmt.Sprintf("app_instance_type=%s", t.config.AppInstanceType),
 		"-var", fmt.Sprintf("agent_instance_count=%d", t.config.AgentInstanceCount),
 		"-var", fmt.Sprintf("agent_instance_type=%s", t.config.AgentInstanceType),
+		"-var", fmt.Sprintf("es_instance=%t", t.config.HasElasticSearch()),
```
Comment (agnivade) [inline deployment/terraform/destroy.go:23]: I think you also need to pass vpc id and everything else (after you remove the default values). This is what I got when I tried to destroy

```
020-06-05T00:04:01.948+0530	info	terraform/engine.go:81	
2020-06-05T00:04:01.986+0530	info	terraform/engine.go:81	var.vpc
2020-06-05T00:04:01.986+0530	info	terraform/engine.go:81	  Enter a value: 
```

While it did go ahead and destroyed everything, it would be good to avoid this.

File: deployment/config.go [hunk 6]
```diff
@@ -77,9 +87,17 @@ func (c *Config) IsValid() error {
 		return fmt.Errorf("load-test package file must be a tar.gz file")
 	}
 
+	if c.HasElasticSearch() && c.VpcID == "" {
```
Comment (agnivade) [inline deployment/config.go:90]: I also got this

```
Error: "domain_name" must start with a lowercase alphabet and be at least 3 and no more than 28 characters long. Valid characters are a-z (lowercase letters), 0-9, and - (hyphen)
```

when I gave "agnivaES" as the name. Might be worth improving the check on L80 to incorporate the length limit and lowercase restriction. 0/5 on whether you want to do it now or in a separate PR.

File: deployment/terraform/output.go [hunk 7]
```diff
@@ -50,6 +50,14 @@ type Output struct {
 			PrivateDNS string `json:"private_dns"`
 		} `json:"value"`
 	} `json:"metricsServer"`
+	ElasticServer struct {
```
Comment (agnivade) [inline deployment/terraform/output.go:53]: Please also show the endpoint info at the end, as part of `(t *Terraform) displayInfo`

File: deployment/terraform/destroy.go (hunk 5 above)
Comment (agnivade) [reply deployment/terraform/destroy.go:23]: Ah wait, will it destroy the entire VPC if I pass the vpc id ? Hmm .. we might need to create a VPC from scratch then. Something for later I guess.

Comment (agnivade) [reply deployment/terraform/destroy.go:23]: Actually, it shouldn't. Because we aren't creating the VPC. Just creating stuff inside it. So should be fine.

File: deployment/config.go [hunk 8]
```diff
@@ -18,19 +18,21 @@ import (
...
-	DBInstanceEngine    string // Type of the DB instance - postgres or mysql.
-	DBUserName          string // Username to connect to the DB.
-	DBPassword          string // Password to connect to the DB.
+	ClusterName           string // Name of the cluster.
+	VpcID                 string // Id of the VPC to be used
+	AppInstanceCount      int    // Number of application instances.
+	AppInstanceType       string // Type of the EC2 instance for app.
+	AgentInstanceCount    int    // Number of agents, first agent and coordinator will share the same instance.
+	AgentInstanceType     string // Type of the EC2 instance for agent.
+	ElasticSearchSettings ElasticSearchSettings
```
Comment (isacikgoz) [inline deployment/config.go:27]: Interesting approach, we were using a flat order here. I think it would be better to keep it consistent. Either group `app`, `agent`, `db` etc. or remove the ES struct here and add individual configurations here. 0/5 on the both TBH.

File: config/deployer.default.json [hunk 9]
```diff
@@ -1,9 +1,17 @@
...
   "AppInstanceCount": 1,
   "AppInstanceType": "c5.xlarge",
   "AgentInstanceCount": 2,
   "AgentInstanceType": "t3.xlarge",
+  "ElasticSearchSettings": {
+    "Enable": false,
+    "InstanceType": "r4.large.elasticsearch",
+    "Version": 6.5,
+    "EBSType": "gp2",
+    "EBSSize": 10
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply config/deployer.default.json:13]: Hum, II agree with moving the type inside but I'd leave the EBS size outside to avoid poking around the terraform code. What do you think?

[the remark this answers is not in the document]
Comment (agnivade) [reply config/deployer.default.json:13]: I think you can just use the method followed for app servers:

```
variable "root_block_device" {
  default = [
    {
      volume_type = "gp2"
      volume_size = 50
    },
  ]
}
```

Just declare the defaults and let terraform use that.

File: deployment/terraform/destroy.go (hunk 5 above)
Comment (ethervoid) [reply deployment/terraform/destroy.go:23]: AFAIK the destroy command only destroys the infrastructure created by terraform so passing the id should be fine and it shouldn't destroy the VPC

File: deployment/config.go (hunk 6 above)
Comment (ethervoid) [reply deployment/config.go:90]: Done in this PR

File: deployment/terraform/output.go (hunk 7 above)
Comment (ethervoid) [reply deployment/terraform/output.go:53]: Done

File: deployment/config.go (hunk 8 above)
Comment (ethervoid) [reply deployment/config.go:27]: Hum, we have the `LoggerSettings`. I think we should have the configuration grouped as we have in the `config.json`. I'd leave it as it is because all the configuration related code is based on that :\. But if @agnivade is agree with you I'll change it

File: config/deployer.default.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (ethervoid) [reply config/deployer.default.json:13]: Done

File: deployment/config.go [hunk 10]
```diff
@@ -77,9 +87,17 @@ func (c *Config) IsValid() error {
...
 	}
 
+	if c.HasElasticSearch() && c.VpcID == "" {
+		return fmt.Errorf("vpc id must be included in order to include an elastic search instance")
+	}
+
 	return nil
 }
 
+func (c *Config) HasElasticSearch() bool {
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/config.go:97]: Removed method :)

File: deployment/terraform/assets/variables.tf [hunk 11]
```diff
@@ -13,6 +17,33 @@ variable "agent_instance_count" {
...
+# Elasticsearch variables
+
+variable "es_instance" {
+  type = bool
+  default = false
+}
+
+variable "es_instance_type" {
+  type = string
+  default = "r4.large.elasticsearch"
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/terraform/assets/variables.tf:29]: Removed the default values and types

File: deployment/terraform/create.go [hunk 12]
```diff
@@ -80,10 +80,16 @@ func (t *Terraform) Create() error {
 
 	err = t.runCommand(nil, "apply",
 		"-var", fmt.Sprintf("cluster_name=%s", t.config.ClusterName),
+		"-var", fmt.Sprintf("vpc=%s", t.config.VpcID),
 		"-var", fmt.Sprintf("app_instance_count=%d", t.config.AppInstanceCount),
 		"-var", fmt.Sprintf("app_instance_type=%s", t.config.AppInstanceType),
 		"-var", fmt.Sprintf("agent_instance_count=%d", t.config.AgentInstanceCount),
 		"-var", fmt.Sprintf("agent_instance_type=%s", t.config.AgentInstanceType),
+		"-var", fmt.Sprintf("es_instance=%t", t.config.HasElasticSearch()),
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/terraform/create.go:88]: Changed

File: deployment/terraform/assets/cluster.tf [hunk 13]
```diff
@@ -419,6 +481,26 @@ resource "aws_security_group" "metrics" {
...
+
+  ingress {
+    from_port       = 443
+    to_port         = 443
+    protocol        = "tcp"
+    security_groups = ["${aws_security_group.app.id}", "${aws_security_group.metrics.id}"]
+  }
+
+  egress {
+    from_port   = 443
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/terraform/assets/cluster.tf:496]: We only want to expose the 443, which is the one is exposing the AWS service. Is not possible to select the port

File: deployment/terraform/metrics.go [hunk 14]
```diff
@@ -102,6 +128,16 @@ func (t *Terraform) setupMetrics(extAgent *ssh.ExtAgent, output *Output) error {
 		return fmt.Errorf("error while uploading dashboard_json: output: %s, error: %w", out, err)
 	}
 
+	if t.config.HasElasticSearch() {
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/terraform/metrics.go:131]: Done

File: deployment/terraform/assets/cluster.tf [hunk 15]
```diff
@@ -147,6 +158,57 @@ resource "aws_instance" "proxy_server" {
...
+          }
+      ]
+  }
+  CONFIG
+
+  depends_on = [
+    aws_iam_service_linked_role.es,
+  ]
+
+  count = var.es_instance ? 1 : 0
```
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/terraform/assets/cluster.tf:208]: Changed

File: deployment/config.go (hunk 2 above)
Comment (ethervoid) [reply deployment/config.go:53]: Done!

File: deployment/config.go (hunk 3 above)
Comment (ethervoid) [reply deployment/config.go:54]: Done

File: deployment/config.go (hunk 8 above)
Comment (agnivade) [reply deployment/config.go:27]: I am fine with grouping. Doesn't make too much of a difference either ways. Let's go with this for now.

File: deployment/terraform/assets/cluster.tf (hunk 13 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply deployment/terraform/assets/cluster.tf:499]: Isn't exposing done with the ingress block ? We are allowing incoming traffic only to 443, which makes sense, but we are also allowing outbound traffic only to 443. I am fine with it, but just curious as to the reason for this.

**Comment by agnivade** [comment]
Any thoughts on this ?

> Lastly, I did not see any uptick in the Query rate panel. Is it because we aren't making any queries at all, or is there a mistake in the panel ? It may well be the first, but just checking.

**Comment by ethervoid** [comment]
Oh sorry I forgot to include the screenshot:

Screenshot from 2020-06-05 19-44-36

I'm using the simple controller and is doing searchs so the grafana panel is working. Which of the controllers are you using?

File: deployment/terraform/assets/cluster.tf (hunk 13 above)
[the remark this answers is not in the document]
Comment (ethervoid) [reply deployment/terraform/assets/cluster.tf:499]: Without egress block set to 443 it wasn't working. Maybe is because some ec2 instances are created in a different vpc?

**Comment by agnivade** [comment]
Ah I was using simulcontroller. Should be fine then. Thanks for checking
